include/PathOrderOptimizer.h

```diff
@@ -365,15 +266,316 @@ class PathOrderOptimizer
+            {
+                if(best_path->is_closed)
+                {
+                    current_position = (*best_path->converted)[best_path->start_vertex]; //We end where we started.
+                }
+                else
+                {
+                    //Pick the other end from where we started.
+                    current_position = best_path->start_vertex == 0 ? best_path->converted->back() : best_path->converted->front();
+                }
+            }
+        }
+
+        return optimized_order;
+    }
+
+    std::vector<OrderablePath> getOptimizerOrderWithConstraints(SparsePointGridInclusive<size_t> line_bucket_grid, size_t snap_radius, const std::unordered_multimap<Path, Path>& order_requirements)
+    {
+        std::vector<OrderablePath> optimized_order; //To store our result in.
+
+        // initialize the roots set with all possible nodes
+        std::unordered_set<Path> roots;
+        std::unordered_set<Path> leaves;
+        for (const auto& path : paths)
+        {
+            roots.insert(path.vertices);
+            leaves.insert(path.vertices);
+        }
+
+        // remove all edges from roots with an incoming edge
+        // result is a set of nodes that have no incoming edges; these are by definition the roots
+        for (const auto& [u, v] : order_requirements)
+        {
+            roots.erase(v);
+            leaves.erase(u);
+        }
+
+        // We used a shared visited set between runs of dfs. This is for the case when we reverse the ordering tree.
+        // In this case two roots can share the same children nodes, but we don't want to print them twice.
+        std::unordered_set<Path> visited;
+        Point current_position = start_point;
+
+        std::function<std::vector<Path>(const Path, const std::unordered_multimap<Path, Path>&)> get_neighbours =
+            [current_position, this](const Path current_node, const std::unordered_multimap<Path, Path>& graph)
+        {
+            std::vector<Path> order; // Output order to traverse neighbors
+
+            const auto& [neighbour_begin, neighbour_end] = graph.equal_range(current_node);
+            auto candidates_iterator = ranges::make_subrange(neighbour_begin, neighbour_end);
+            std::unordered_set<Path> candidates;
+            for (const auto& [_, neighbour] : candidates_iterator)
+            {
+                candidates.insert(neighbour);
+            }
+
+            auto local_current_position = current_position;
+            while (candidates.size() != 0)
+            {
+                Path best_candidate = findClosestPathVertices(local_current_position, candidates);
+
+                candidates.erase(best_candidate);
+                order.push_back(best_candidate);
+
+                // update local_current_position
+                for (auto& path : paths)
+                {
+                    if (path.vertices == best_candidate)
+                    {
```

Comment:
Didn't we have a map from one type of path ordering to another type of path ordering so that we don't have a `O(n)` slow search to find the correct corresponding path type?